text/nodejs/0014-yarn-berry.md

```diff
@@ -0,0 +1,280 @@
+
+  [[order.group]]
+    id = "paketo-buildpacks/image-labels"
+    optional = true
+
+...
+
+```
+
+The `yarn-berry-install` group is placed before the traditional `yarn-install`
+group since it detects on the presence of `.yarnrc.yml` and both app types
+will contain `yarn.lock` files.
+
+### Detection
+
+- `.yarnrc.yml` was created with the intention of making it easy to detect
+  whether a project uses Yarn Classic or Berry (Berry ignores `.yarnrc`
+  completely), and should be used for the detection of `yarn-berry-install`.
+
+### Build
+
+Yarn Berry includes a number of features which seek to remove the necessity of
+`yarn install`. As such, the buildpack should allow users to take advantage of
+these features by avoiding invocations of `yarn install` when it is reasonable
+to do so. For those who opt out of "Plug 'n Play", Berry still supports using
+`node_modules`.
+
+The buildpack should run `yarn install` in the following cases:
+
+- If an application uses `node_modules`, which may be determined by the
+  presence of a `node_modules` directory or through `.yarnrc.yml` configuration.
+
+- If the `yarn.lock` has changed when using `node_modules`.
+
+- If there is no local cache (`.yarn/cache` by default, but can be configured) in
+  the working directory, whether `.pnp.cjs` exists or not.
+
+- If there is no `.pnp.cjs` in the working directory, even if a local cache exists.
+
+The buildpack should NOT run `yarn install` when:
+
+- Both a `.pnp.cjs` file and a local cache are present in the working
+  directory. The buildpack should assume that these two artifacts indicate a
+  desire to leverage Plug'n'Play and should not need to run `yarn install`
+  whether or not the `yarn.lock` has changed between builds.
+
+
+The buildpack will store the Yarn cache in a layer and configure Yarn to
+reference this layer, ideally via the `YARN_CACHE_FOLDER` environment variable.
+For Berry projects using `node_modules`, the behaviour will largely mimic the
+existing Yarn Install buildpack, with accommodations for the CLI changes.
+
+The buildpack will generate a filesystem-based SBoM as `yarn-install` does currently.
+
+### Layer Reuse 
+
+The Yarn Install buildpack currently uses a checksum of the output from `yarn
+config list` and the contents of `yarn.lock` to decide whether to reuse package layers.
+`yarn config list` has been removed in Berry, but the new `yarn info` command, which
+provides information about the project's direct and transitive dependencies,
+should work well as a replacement.
+
+### `--frozen-lockfile`
+
+The existing Yarn Install buildpack passes the `--frozen-lockfile` argument to
+every `yarn install` call, which will not update the `yarn.lock` file, and will
+fail if the `yarn.lock` file needs to be updated. As such, developers must run
+`yarn install` locally in order to update their app's `yarn.lock` file.
+`--frozen-lockfile` has been deprecated in favour of `--immutable`, which this
+buildpack will utilize.
```

Comment:
I am bit concerned by this.
I agree `--immutable` is a good idea in theory.
But I had strange cases where for an identical node and yarn version, `yarn install --immutable` would fail on some env and succeed on others. Very frustrating and hard to debug.

Should we instead not specify the option by default and let users either
- add `enableImmutableInstalls: false` to `.yarnrc.yml`
- or set with yarn env vars mechanism: `YARN_ENABLE_IMMUTABLE_INSTALLS`